modules/indexing/src/test/java/org/apache/ignite/internal/processors/cache/PrecisionTest.java

```diff
@@ -0,0 +1,344 @@
+    public void testClientInsertValueVarColumns() throws Exception {
+        try (IgniteClient client = Ignition.startClient(new ClientConfiguration().setAddresses("127.0.0.1"))) {
+            ClientCache cc = client.cache("PERSON");
+
+            StringBuilder validStr = new StringBuilder("12345");
+
+            SqlFieldsQuery strQry = new SqlFieldsQuery("insert into PERSON(_KEY, name) values (?, ?)");
+            SqlFieldsQuery binQry = new SqlFieldsQuery("insert into PERSON(_KEY, arr) values (?, ?)");
+
+            for (int i = 1; i < CNT; i++) {
+                String v = validStr.append("0").toString();
+                int len = 5 + i;
+
+                assertClientPrecision(new Callable<Object>() {
+                    @Override public Object call() {
+                        cc.put(KEY, new Person(v));
+
+                        return null;
+                    } }, "NAME", len);
+
+                assertClientPrecision(new Callable<Object>() {
+                    @Override public Object call() {
+                        cc.put(KEY, new Person(v.getBytes(StandardCharsets.UTF_8)));
+
+                        return null;
+                    } }, "ARR", len);
+
+                assertClientPrecision(new Callable<Object>() {
+                    @Override public Object call() {
+                        cc.query(strQry.setArgs(KEY, v)).getAll();
+
+                        return null;
+                    } }, "NAME", len);
+
+                assertClientPrecision(new Callable<Object>() {
+                    @Override public Object call() {
+                        cc.query(binQry.setArgs(KEY, v.getBytes(StandardCharsets.UTF_8))).getAll();
+
+                        return null;
+                    } }, "ARR", len);
+            }
+        }
+    }
+
+    /** */
+    private void assertPrecision(Callable<Object> clo, String colName, int len) {
+        GridTestUtils.assertThrows(null, clo, CacheException.class,
+            "Value for a column '" + colName + "' is too long. Maximum length: 5, actual length: " + len);
+
+        assertNull(cache.get(0));
+    }
+
+    /** */
+    private void assertClientPrecision(Callable<Object> clo, String colName, int len) {
+        GridTestUtils.assertThrows(null, clo, ClientException.class,
+            "Value for a column '" + colName + "' is too long. Maximum length: 5, actual length: " + len);
+
+        assertNull(cache.get(0));
+    }
+
+    /** */
+    static class Person {
+        /** */
+        @QuerySqlField(precision = 5)
+        private final String name;
+
+        @QuerySqlField(precision = 5)
+        private final byte[] arr;
+
+        /** */
+        Person(String name) {
+            this.name = name;
+            this.arr = null;
+        }
+
+        /** */
+        Person(byte[] arr) {
+            this.name = null;
+            this.arr = arr;
+        }
+
+        /** {@inheritDoc} */
+        @Override public int hashCode() {
+            if (name != null)
```

Comment:
Where do you use `equals` and `hashCode`?

Reply:
Will remove it.